inference/__init__.py

```diff
@@ -0,0 +1,492 @@
+        self.beta1 = spqr_host.beta1
+        self.beta2 = spqr_host.beta2
+        self.buff0 = spqr_host.buff0
+        self.row_offsets = spqr_host.row_offsets
+        self.col_vals = spqr_host.col_vals
+        self.deq_w = None
+        self.in_perm = spqr_host.in_perm
+        self.out_perm = spqr_host.out_perm
+
+
+    def to_device(self, device: torch.device):
+        self.buff0 = self.buff0.to(device=device)
+        self.row_offsets = self.row_offsets.to(device=device)
+        self.col_vals = self.col_vals.to(device=device)
+
+    @property
+    def nnz(self):
+        return self.col_vals.shape[0]
+
+    def _apply(self, fn):
+        # Apply the function to all parameters and buffers
+        super(SPQRModule, self)._apply(fn)
+
+        # Apply the function to custom attributes
+        self.buff0 = fn(self.buff0)
+        self.row_offsets = fn(self.row_offsets)
+        self.col_vals = fn(self.col_vals)
+
+        if self.in_perm is not None:
+            self.in_perm = fn(self.in_perm)
+
+        if self.out_perm is not None:
+            self.out_perm = fn(self.out_perm)
+
+
+        return self
+
+    @property
+    def density(self) -> float:
+        return self.col_vals.shape[0] / (self.m * self.n)
+
+    @property
+    def sparsity(self) -> float:
+        return 1 - self.density
+
+    def forward(self, x: T) -> T:
+        inner_dim = x.shape[1]
+
+        y = torch.empty((inner_dim * self.m), dtype=torch.float16, device=self.buff0.device)
+
+        for i in range(inner_dim):
+            _x = x[..., i, :].flatten()
+            if self.in_perm is not None:
+                _x = _x[self.in_perm]
+            spqr_cuda.spqr_mul(
+                self.m,
+                self.n,
+                self.bits,
+                self.beta1,
+                self.beta2,
+                self.buff0,
+                self.row_offsets,
+                self.col_vals,
+                self.nnz,
+                _x,
+                y[(i * self.m):((i + 1) * self.m)],
+                FeatureFlag.SPARSE_FUSED_FP32)
+
+        return y.reshape((1, inner_dim, self.m))
+
+
+# Compression
+
+def compress(W, bit_count):
+    W = list_flatten(W)
+    value_count = W.shape[0]
+
+    # Values per bucket
+    values_per_bucket = 64 // bit_count
+
+    total_buckets = (value_count + values_per_bucket - 1) // values_per_bucket
+    out = torch.zeros((total_buckets), dtype=torch.int64)
+
+    # Position within bucket
+    buff_id = 0
+
+    # Current bucket
+    bucket_id = 0
+
+    # Current bucket
+    bucket = np.uint64(0)
+
+    for i in range(value_count):
+        bucket |= np.uint64(W[i]) << np.uint64(buff_id * bit_count)
+        buff_id += 1
+
+        if buff_id == values_per_bucket:
+            out[bucket_id] = bucket
+            buff_id = 0
+            bucket = np.uint64(0)
+            bucket_id += 1
+
+    if buff_id > 0:
+        out[bucket_id] = bucket
+
+    return out
+
+
+def _spqr_dequantize(p: SPQRModule, nnz):
```

Comment:
What is nnz here?

Reply:
Refactored into an internal private method.